# 판다스 실습 노트북


- Pandas 기본 문법을 확인하는 4개 섹션으로 구성된 연습문제
- 각 섹션은 문제 3개로 구성됨
- 사용 데이터 파일: `semicond_pandas_practice.csv`


## 준비

실습 노트북과 `semicond_pandas_practice.csv` 파일을 같은 폴더에 둔 뒤 진행하세요.

# Section 1. 데이터프레임 기본 확인과 데이터 선택

## 문제 1. CSV 파일 불러오기와 기본 확인

`semicond_pandas_practice.csv` 파일을 읽어 `df`라는 데이터프레임에 저장하세요.

다음 내용을 확인하세요.

- 상위 5개 행 출력
- 데이터 크기 출력
- 컬럼별 자료형 출력
- 수치형/문자형 컬럼의 기초 통계량 출력

In [1]:
# 파일 읽기
import pandas as pd

df = pd.read_csv("semicond_pandas_practice.csv")

In [2]:
# 상위 5개 행 출력
df.head()

,Product_ID,Wafer_ID,Lot_ID,Factory,Line,Shift,Process_Mode,Product_Type,Material_Supplier,Chamber_ID,Temperature,Chamber_Pressure,Gas_Flow,Etch_Rate,Chamber_Humidity,Particle_Count,Thickness_Uniformity,Voltage_Stability,Cycle_Time,Defect
0,CHIP-00001,WAFER-00001,LOT-027,A,L3,Night,High_Speed,Standard,S3,CH-04,497.51,2.497,117.99,1.408,38.19,19.7,0.9311,0.9828,30.19,0
1,CHIP-00002,WAFER-00002,LOT-024,C,L1,Night,Normal,Advanced,S1,CH-04,501.13,2.619,NaN,1.536,44.54,27.9,0.9162,0.9394,31.93,0
2,CHIP-00003,WAFER-00003,LOT-002,B,L1,Day,High_Speed,Advanced,NaN,CH-04,484.87,2.824,113.12,1.584,NaN,18.6,0.9379,NaN,37.58,1
3,CHIP-00004,WAFER-00004,LOT-007,B,L2,Day,Normal,Premium,S1,CH-03,488.05,2.495,116.33,1.564,35.40,26.8,0.9781,0.9615,26.08,0
4,CHIP-00005,WAFER-00005,LOT-004,A,L1,Night,High_Speed,Standard,S3,CH-04,474.32,2.452,127.21,1.542,50.63,27.4,0.9283,NaN,29.44,0


In [3]:
# 상위 5개 행 출력
df.shape

(3600, 20)

In [4]:
# 컬럼별 자료형 출력
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 3600 entries, 0 to 3599
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Product_ID            3600 non-null   str    
 1   Wafer_ID              3600 non-null   str    
 2   Lot_ID                3600 non-null   str    
 3   Factory               3600 non-null   str    
 4   Line                  3600 non-null   str    
 5   Shift                 3600 non-null   str    
 6   Process_Mode          3600 non-null   str    
 7   Product_Type          3600 non-null   str    
 8   Material_Supplier     3515 non-null   str    
 9   Chamber_ID            3600 non-null   str    
 10  Temperature           3600 non-null   float64
 11  Chamber_Pressure      3471 non-null   float64
 12  Gas_Flow              3448 non-null   float64
 13  Etch_Rate             3600 non-null   float64
 14  Chamber_Humidity      3502 non-null   float64
 15  Particle_Count        3600 non-n

In [5]:
# 수치형/문자형 컬럼의 기초 통계량 출력
display(df.describe())

display(df.describe(include='object'))

,Temperature,Chamber_Pressure,Gas_Flow,Etch_Rate,Chamber_Humidity,Particle_Count,Thickness_Uniformity,Voltage_Stability,Cycle_Time,Defect
count,3600.000000,3471.000000,3448.000000,3600.000000,3502.000000,3600.000000,3466.000000,3444.000000,3600.000000,3600.000000
mean,481.788789,2.488810,119.744014,1.545121,37.895357,20.150500,0.934249,0.970783,31.017936,0.066667
std,12.547289,0.201631,7.413984,0.098587,5.478124,7.052129,0.022771,0.012122,6.480908,0.249478
min,428.750000,1.527000,93.020000,1.189000,20.270000,2.900000,0.858000,0.932400,20.510000,0.000000
25%,473.430000,2.358000,114.967500,1.479000,34.200000,16.300000,0.919400,0.962400,28.150000,0.000000
50%,481.930000,2.487000,119.690000,1.548000,37.930000,19.600000,0.934400,0.971100,30.675000,0.000000
75%,490.147500,2.624000,124.440000,1.607000,41.427500,22.800000,0.950300,0.979400,32.872500,0.000000
max,520.170000,3.414000,141.640000,1.906000,54.040000,95.000000,0.990000,1.000000,115.580000,1.000000


C:\Users\metam\AppData\Local\Temp\ipykernel_1052\394665998.py:4: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  display(df.describe(include='object'))


,Product_ID,Wafer_ID,Lot_ID,Factory,Line,Shift,Process_Mode,Product_Type,Material_Supplier,Chamber_ID
count,3600,3600,3600,3600,3600,3600,3600,3600,3515,3600
unique,3600,3600,40,3,3,2,2,3,3,4
top,CHIP-00001,WAFER-00001,LOT-017,A,L2,Day,Normal,Advanced,S1,CH-01
freq,1,1,128,1298,1302,2206,2582,1381,1680,1019


## 문제 2. 컬럼 선택, `loc`, `iloc` 연습

다음 작업을 수행하세요.

- `Product_ID`, `Factory`, `Line`, `Shift`, `Product_Type`, `Defect` 컬럼만 선택해 `basic_df`에 저장하고 첫 3행을 확인하세요.
- `bisic_df`에서 `loc`를 사용해 0, 1, 2번 행의 `Factory`, `Product_Type`, `Defect` 컬럼을 선택하세요.
-  `bisic_df`에서 `iloc`를 사용해 처음 3개 행과 처음 5개 컬럼을 선택하세요.

In [6]:
# 6개 컬럼 선택하여 basic_df 만들기
basic_df = df[["Product_ID", "Factory", "Line", "Shift", "Product_Type", "Defect"]]
basic_df.head(3)

,Product_ID,Factory,Line,Shift,Product_Type,Defect
0,CHIP-00001,A,L3,Night,Standard,0
1,CHIP-00002,C,L1,Night,Advanced,0
2,CHIP-00003,B,L1,Day,Advanced,1


In [7]:
# loc 사용
basic_df.loc[[0, 1, 2], ["Factory", "Product_Type", "Defect"]]


,Factory,Product_Type,Defect
0,A,Standard,0
1,C,Advanced,0
2,B,Advanced,1


In [8]:
# iloc 사용
basic_df.iloc[:3, :5]

,Product_ID,Factory,Line,Shift,Product_Type
0,CHIP-00001,A,L3,Night,Standard
1,CHIP-00002,C,L1,Night,Advanced
2,CHIP-00003,B,L1,Day,Advanced


## 문제 3. 조건에 맞는 데이터 필터링

다음 조건에 해당하는 데이터를 `df`에서 각각 필터링하고, 데이터 개수도 함께 출력하세요.

- 불량 제품만 선택하여 `defect_df`에 대입하세요.
- `Factory`가 `C`이고 `Product_Type`이 `Premium`인 제품만 선택하여 `c_premium_df`에 대입하세요.
- 각 데이터프레임의 샘플 수를 출력하세요. 각각 240, 493로 나와야함

In [9]:
defect_df = df[df["Defect"] == 1]
c_premium_df = df[(df["Factory"] == "C") & (df["Product_Type"] == "Premium")]

print("불량 제품 수:", len(defect_df))
print("Factory C & Premium 제품 수:", len(c_premium_df))

불량 제품 수: 240
Factory C & Premium 제품 수: 493


---

# Section 2. 통계, 파생변수, 정렬, 결측치, 이상치 처리

## 문제 4. 기본 통계와 파생변수 만들기

다음 작업을 `df`에 대해 수행하세요.

- 전체 불량률을 계산하세요. 불량률은 얼마인가요? (6.67%)
- `Particle_Count / Cycle_Time`을 계산하여 `Particle_per_Cycle` 컬럼을 추가하세요.
- `Particle_per_Cycle` 값이 큰 순서로 상위 3개 행을 확인하세요. (인덱스 번호 2949, 2919, 1747이 나와야함)

In [10]:
defect_rate = df["Defect"].mean()
print(f"전체 불량률: {defect_rate:.2%}")

df["Particle_per_Cycle"] = df["Particle_Count"] / df["Cycle_Time"]
df.sort_values("Particle_per_Cycle", ascending=False).head(3)


전체 불량률: 6.67%


,Product_ID,Wafer_ID,Lot_ID,Factory,Line,Shift,Process_Mode,Product_Type,Material_Supplier,Chamber_ID,...,Chamber_Pressure,Gas_Flow,Etch_Rate,Chamber_Humidity,Particle_Count,Thickness_Uniformity,Voltage_Stability,Cycle_Time,Defect,Particle_per_Cycle
2949,CHIP-02950,WAFER-02950,LOT-040,B,L2,Day,Normal,Standard,S2,CH-01,...,2.685,119.38,1.402,47.62,74.5,0.9241,1.0000,21.23,1,3.509185
2919,CHIP-02920,WAFER-02920,LOT-011,C,L2,Day,Normal,Premium,S1,CH-03,...,2.516,120.75,1.466,38.17,93.5,0.9271,0.9629,30.22,1,3.093977
1747,CHIP-01748,WAFER-01748,LOT-011,C,L1,Day,High_Speed,Premium,S1,CH-03,...,2.516,120.75,1.466,38.17,93.5,0.9271,0.9629,30.22,1,3.093977


## 문제 5. 결측치 처리

`df`에 결측치를 처리하기 위해 `df_clean`으로 데이터를 복했습니다. `df_clean`에 대해서 결측치가 있는 컬럼을 확인한 뒤, 다음 방식으로 결측치를 처리하세요. 수치형 컬럼은 `num_cols`, 문자형 컬럼은 `cat_cols`를 사용하세요.

- 전체 결측치의 개수를 확인하세요. (754개)
- 수치형 컬럼의 결측치는 각 컬럼의 중앙값으로 채우세요.
- 문자형 컬럼의 결측치는 `"Unknown"`으로 채우세요.
- 결측치 처리 후 `df_clean`에 결측치가 남아 있는지 다시 확인하세요.

In [11]:
df_clean = df.copy()
num_cols = df_clean.select_dtypes(include="number").columns
cat_cols = df_clean.select_dtypes(exclude="number").columns

print("결측치 총합:", df.isnull().sum().sum())

df_clean[num_cols] = df_clean[num_cols].fillna(df_clean[num_cols].median())
df_clean[cat_cols] = df_clean[cat_cols].fillna("Unknown")

print("남은 결측치 총합:", df_clean.isna().sum().sum())


결측치 총합: 754
남은 결측치 총합: 0


## 문제 6. IQR 방식으로 이상치 찾기와 제거하기

`df_clean`에 대해서 `Particle_Count` 컬럼을 대상으로 IQR 방식의 이상치 기준을 계산하세요.

다음 값을 구하세요.

- Q1, Q3, IQR, lower bound, upper bound

그 다음 다음 작업을 수행하세요.

- `Particle_Count`가 IQR 기준을 벗어난 행을 선택해서 `outlier_df`에 대입하고 행수를 출력하세요. 61행

- 이상치를 제거한 결과를 `df_no_outlier`에 저장하고 행수를 확인하세요. `no_outlier_df`의 행수는 3539가 되야함

In [12]:
col = "Particle_Count"

# qauntail() 함수를 사용해서 Q1, Q3, IQR를 구하세요.
Q1 = df_clean[col].quantile(0.25)
Q3 = df_clean[col].quantile(0.75)
IQR = Q3 - Q1
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

print("Q1:", Q1)
print("Q3:", Q3)
print("IQR:", IQR)
print("lower bound:", lower_bound)
print("upper bound:", upper_bound)

# Particle_Count가 IQR 기준을 벗어난 행만 필터링
outlier_cond = (df_clean[col] < lower_bound) | (df_clean[col] > upper_bound)
outlier_df = df_clean[outlier_cond]
print("이상치 포함 데이터 크기:", outlier_df.shape)


no_outlier_df = df_clean[~outlier_cond].copy()
print("이상치 제거 후 데이터 크기:", no_outlier_df.shape)


Q1: 16.3
Q3: 22.8
IQR: 6.5
lower bound: 6.550000000000001
upper bound: 32.55
이상치 포함 데이터 크기: (61, 21)
이상치 제거 후 데이터 크기: (3539, 21)


---

# Section 3. 그룹화, 피벗테이블, 조건부 불량률

## 문제 7. 공장별 불량률 계산

`df_clean`에 대해서 공장별(`Factory`)로 다음 값을 계산하세요. hint: groupby해서 size, sum, mean을 사용하세요.

- 생산 건수
- 불량 건수
- 불량률

계산한 정보를 담은 `df_factory` 데이터프레임에 대해서 불량률이 높은 순서로 정렬하고 어떤 공장의 불량률이 가장 높은지 확인하세요. C공장

## 문제 8. 공장 × 교대조, 공장별 생산건수 계산

'df_clean'에 대해 다음 작업을 수행하세요.

- `pivot_table()`을 사용하여 공장(Factory) × 교대조별(Shift) 불량률을 계산하여 `f_shift_rate`에 저장하고 확인하세요.
- `crosstab()`을 사용하여 공장별 (Product_Type) 생산 건수를 `f_product_count`에 저장하고 확인하세요.

## 문제 9. 조건부확률 기반 생산 배정 의사결정

Premium 제품 10,000개를 추가 생산해야 한다고 가정합니다.

`df_clean`을 사용하여 공장 A, B, C 중 어느 공장에 배정하는 것이 가장 합리적인지 판단하세요.

단, 전체 공장별 불량률만 보지 말고, `Product_Type = Premium` 조건에서의 공장별 불량률을 기준으로 판단하세요.


---

# Section 4. 데이터 시각화

## 문제 10. 전체 양품/불량 분포 시각화

`df_clean`을 사용하여 Defect 컬럼의 양품과 불량 제품의 개수를 막대그래프로 시각화하세요.

## 문제 11. 공장별 불량률 막대그래프

`Factory`별 불량률을 계산한 뒤, 불량률이 높은 순서대로 막대그래프로 시각화하세요.

## 문제 12. Particle_Count와 불량 여부의 관계 시각화

`Particle_Count`가 불량 여부와 관련이 있는지 확인하기 위해 `Defect`별 `Particle_Count` 분포를 박스플롯으로 시각화하세요.